governance/contract_connector.go

```diff
@@ -0,0 +1,155 @@
+	tx, evm, err := c.prepareCall(num)
+	if err != nil {
+		return nil, err
+	}
+
+	// ignore error; if error, all nodes will have the same error
+	res, _ := c.callTx(tx, evm)
+
+	// ignore error; if error, all nodes will have the same error
+	pset, err := c.parse(res)
+	if err != nil {
+		return params.NewGovParamSet(), nil
+	}
+	return pset, nil
+}
+
+func (c *contractCaller) prepareCall(num *big.Int) (*types.Transaction, *vm.EVM, error) {
+	tx, err := c.makeTx(govParamAbi, govparamfunc, num)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	evm, err := c.makeEVM(tx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return tx, evm, nil
+}
+
+// Make contract execution transaction
+func (c *contractCaller) makeTx(contractAbi abi.ABI, fn string, args ...interface{},
+) (*types.Transaction, error) {
+	calldata, err := contractAbi.Pack(fn, args...)
+	if err != nil {
+		logger.Error("Could not pack ABI", "err", err)
+		return nil, err
+	}
+
+	rules := c.chain.Config().Rules(c.chain.CurrentHeader().Number)
+	intrinsicGas, err := types.IntrinsicGas(calldata, nil, false, rules)
+	if err != nil {
+		logger.Error("Could not fetch intrinsicGas", "err", err)
+		return nil, err
+	}
+
+	var (
+		from       = common.Address{}
+		to         = &c.contractAddr
+		nonce      = uint64(0)
+		amount     = big.NewInt(0)
+		gasLimit   = uint64(1e8)
+		gasPrice   = big.NewInt(0)
+		checkNonce = false
+	)
+
+	tx := types.NewMessage(from, to, nonce, amount, gasLimit, gasPrice, calldata,
+		checkNonce, intrinsicGas)
+	return tx, nil
+}
+
+// Make contract execution transaction
+func (c *contractCaller) makeEVM(tx *types.Transaction) (*vm.EVM, error) {
+	// Load the latest state
+	block := c.chain.GetBlockByNumber(c.chain.CurrentHeader().Number.Uint64())
+	if block == nil {
+		logger.Error("Could not find the latest block", "num", c.chain.CurrentHeader().Number.Uint64())
+		return nil, errors.New("no block")
+	}
+
+	statedb, err := c.chain.StateAt(block.Root())
+	if err != nil {
+		logger.Error("Could not find the state", "err", err, "num", c.chain.CurrentHeader().Number.Uint64())
+		return nil, err
+	}
+
+	// Run EVM at given states
+	evmCtx := blockchain.NewEVMContext(tx, block.Header(), c.chain, nil)
+	// EVM demands the sender to have enough KLAY balance (gasPrice * gasLimit) in buyGas()
+	// After KIP-71, gasPrice is baseFee (=nonzero), regardless of the msg.gasPrice (=zero)
+	// But our sender (0x0) won't have enough balance. Instead we override gasPrice = 0 here
+	evmCtx.GasPrice = big.NewInt(0)
+	evm := vm.NewEVM(evmCtx, statedb, c.chain.Config(), &vm.Config{})
+	return evm, nil
+}
+
+// Execute contract call at the latest block context
+func (c *contractCaller) callTx(tx *types.Transaction, evm *vm.EVM) ([]byte, error) {
+	res, _, kerr := blockchain.ApplyMessage(evm, tx)
+	if kerr.ErrTxInvalid != nil {
+		logger.Warn("Invalid tx")
+		return nil, kerr.ErrTxInvalid
+	}
+
+	return res, nil
+}
+
+func (c *contractCaller) parse(b []byte) (*params.GovParamSet, error) {
```

Comment:
@ian0371 Please use a specific name since there could be more `parseX` functions (though there isn't now)
```suggestion
func (c *contractCaller) parseGetAllParamsAt(b []byte) (*params.GovParamSet, error) {
```

Reply:
fixed in f8a0379